ragstack/colbert/cassandra_retriever.py

```diff
@@ -94,17 +137,181 @@ def __init__(
         Closes any open resources held by the retriever.
         """
-        pass
+        self.executor.shutdown(wait=True)
+
+    def _query_relevant_chunks(
+        self, query_encodings: Tensor, top_k: int, query_timeout: int
+    ) -> Set[Tuple[str, int]]:
+        """
+        Retrieves the top_k ANN results for each embedded query token.
+        """
+        chunks: Set[Tuple[str, int]] = set()
+
+        def query_and_process(query_vector: Tensor) -> Set[Tuple[str, int]]:
+            local_chunks = set()
+            try:
+                async_future = self.vector_store.session.execute_async(
+                    self.vector_store.query_colbert_ann_stmt,
+                    [list(query_vector), top_k],
+                    timeout=query_timeout,
+                )
+                embeddings = async_future.result()
+                for embedding in embeddings:
+                    local_chunks.add((embedding.doc_id, embedding.chunk_id))
+            except ReadTimeout:
+                logging.warn(f"Query timeout with params: {query_vector}, {top_k}")
+                # Handle the timeout or other potential exceptions as needed
+            except Exception as e:
+                logging.error(
+                    f"Error during query execution or result fetching: {str(e)}"
+                )
+            return local_chunks
+
+        executor_futures = {
+            self.executor.submit(query_and_process, qv): qv for qv in query_encodings
+        }
+        for executor_future in as_completed(executor_futures):
+            query_vector = executor_futures[executor_future]
+            try:
+                result_chunks = executor_future.result()
+                chunks.update(result_chunks)
+            except Exception as exc:
+                logging.error(
+                    f"query_vector: {query_vector} generated an exception: {exc}"
+                )
+
+        return chunks
+
+    def _retrieve_chunks(
+        self, chunks: Set[Tuple[str, int]], query_timeout: int
+    ) -> Dict[Tuple[str, int], List[Tensor]]:
+        """
+        Retrieves chunk data for a set of doc_id and chunk_id pairs, returning a dictionary mapping each pair to a list of PyTorch tensors.
+        """
+        chunk_data: Dict[Tuple[str, int], List[Tensor]] = {}
+
+        def fetch_chunk_data(
+            doc_id: str, chunk_id: int
+        ) -> Tuple[Tuple[str, int], List[Tensor]]:
+            try:
+                async_future = self.vector_store.session.execute_async(
+                    self.vector_store.query_colbert_chunks_stmt,
+                    [doc_id, chunk_id],
+                    timeout=query_timeout,
+                )
+                rows = async_future.result()
+                return (doc_id, chunk_id), [
+                    torch.tensor(row.bert_embedding) for row in rows
+                ]
+            except ReadTimeout:
+                logging.warn(f"Query timeout for doc_id {doc_id}, chunk_id {chunk_id}")
```

Comment:
shall we retry? 
In general I believe if one query fails, we should raise the error to the caller

Reply:
i'll change this to error level for now.